pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -314,11 +314,37 @@ func (c *Cluster) resolveNode(nodeName, deviceClass string) *cephv1.Node {
 	if rookNode == nil {
 		return nil
 	}
-	rookNode.Resources = k8sutil.MergeResourceRequirements(rookNode.Resources, cephv1.GetOSDResources(c.spec.Resources, deviceClass))
```

Comment:
I'd suggest we put this line back to get the resources for the node. Then we add an override section like this:
```go
if rookNode.ResourceClassName != "" {
  // Override the resources for the node
}
```

This method is also only for determining the resources for all OSDs on a node. We need to find a different place to override the resources for a specific device on that node if it has a different `resourceClassName`.
The challenge is that the operator doesn't know the device name, only the osd prepare job knows the device name. The device names may also change at a later time, and then we would lose the mapping to the resourceDeviceName. I'm thinking the best way to support device-level resources is really with the device classes, which is already supported. I think this requested feature only can be supported at the `node` level, or the `storageClassDeviceSet` level, but then is it worth it? 

@dungdm93 Would it work for you to use the deviceClass for specific resources as in #8007? I'm not sure the requested granularity will work well because of changing device names.

Reply:
`changing device names` means that OSDs that are created may change their devices if the operator reconciles?

Still, we can have a ResourceClass mapping added to support the easier way of applying the resources over the nodes and devices, what are your thoughts on this? 

But I see with the current changes node's resources specs are given higher priority than the device's resources specs, which should be the opposite. https://github.com/rook/rook/blob/4ec496dd478a90c51814bccf9ef778fb7aed7b17/pkg/operator/k8sutil/resources.go#L150

Reply:
>The challenge is that the operator doesn't know the device name, only the osd prepare job knows the device name.
 The device names may also change at a later time, and then we would lose the mapping to the resourceDeviceName.

@travisn, I am still not clear with the thought that how `devices change` affect the mappings,
Like we can iterate over the node's devices and get ResourceClass for underlining devices present on OSDs, which probably happens if it reconciles.

Reply:
Let's set aside the question of device names changing for a moment... At this point in the code, the operator is generating the pod spec for the OSD daemon. The operator does not know the name of the device on which the OSD was created. The operator only knows which node the OSD came from. So we have no way to map the resources to which OSD it's desired to apply them to. Since we only know other properties like its device class, my thought is that we really just rely on the device class for those resource assignments, which capability we already have today.

Reply:
@travisn, Thanks Travis for the explanation, I got your point of the operator not having knowledge of node device class, I agree with you on going with the cluster level device-class configuration.

And wanted to ask do we need a ResourceClass for applying the limits on node and devices( device-level ), or it's not that worth having it?

And want to understand how osd-prepare job help in creating osd can you suggest any doc, 
looking at the code I understood that the operator goes and creates awaitingStatusConfigMaps and the prepare jobs,  and the reconciler has the further logic for the jobs for creating osds.